CONTRIBUTING.md

```diff
@@ -422,25 +422,36 @@ Note: If you have run `test/ipynb/mpl_tester.ipynb` locally it is possible some
 
 ## Style and lint
 
-Qiskit Terra uses 2 tools for verify code formatting and lint checking. The
+Qiskit Terra uses three tools for verify code formatting and lint checking. The
 first tool is [black](https://github.com/psf/black) which is a code formatting
 tool that will automatically update the code formatting to a consistent style.
 The second tool is [pylint](https://www.pylint.org/) which is a code linter
 which does a deeper analysis of the Python code to find both style issues and
-potential bugs and other common issues in Python.
-
-You can check that your local modifications conform to the style rules
-by running `tox -elint` which will run `black` and `pylint` to check the local
-code formatting and lint. If black returns a code formatting error you can
-run `tox -eblack` to automatically update the code formatting to conform to
-the style. However, if `pylint` returns any error you will have to fix these
-issues by manually updating your code.
-
-Because `pylint` analysis can be slow, there is also a `tox -elint-incr` target, which only applies
-`pylint` to files which have changed from the source github. On rare occasions this will miss some
-issues that would have been caught by checking the complete source tree, but makes up for this by
-being much faster (and those rare oversights will still be caught by the CI after you open a pull
-request).
+potential bugs and other common issues in Python. The third tool is the linter
+[ruff](https://github.com/charliermarsh/ruff), which has been recently
+introduced into Qiskit Terra on an experimental basis. Only a very small number
+of rules are enabled.
+
+You can check that your local modifications conform to the style rules by
+running `tox -elint` which will run `black`, `ruff`, and `pylint` to check the
+local code formatting and lint. If black returns a code formatting error you can
+run `tox -eblack` to automatically update the code formatting to conform to the
+style. However, if `ruff` or `pylint` return any error you will have to fix
+these issues by manually updating your code.
+
+Because `pylint` analysis can be slow, there is also a `tox -elint-incr` target,
+which runs `black` and `ruff` just as `tox -elint` does, but only applies
+`pylint` to files which have changed from the source github. On rare occasions
+this will miss some issues that would have been caught by checking the complete
+source tree, but makes up for this by being much faster (and those rare
+oversights will still be caught by the CI after you open a pull request).
+
+Because they are so fast, it is sometimes convenient to run the tools `black` and `ruff` separately
+rather than via `tox`. If you have installed the development packages in your python environment via
+`pip install -r requirements.dev`, then `ruff` and `black` will be available. Otherwise install them
+with `pip install ruff black`. Running the commands `ruff check qiskit test tools examples setup.py`
```

Comment:
This is potentially problematic to suggest because of our pinning. If you run `pip install black` for example it will pull in a `23.x.y` version which will behave differently than what CI will check (I think we can look at bumping it in a separate PR as it's separate), but it might be good to include the pins here although I guess it's one more place to keep in sync.

Reply:
I'd recommend we remove this paragraph for the exact problem you're describing.

One option is to have multiple environments:

- `tox -e fmt`
- `tox -e fix`
- `tox -e lint`
- `tox -e ruff`
- `tox -e black`
- `tox -e pylint`

The downside is that `tox.ini` will be more awkward, but that doesn't seem like a big deal to me. Optimize for developer experience, not how elegant a config file is that very few people touch.

Reply:
> Won't tox -e ruff be much slower than just running ruff?

Why? It will be annoying the first time you have to install the virtual environment. But otherwise it should be roughly the same speed.

Reply:
I'm looking for a good place in the contributors doc to warn users that they should install dev tools only via `requirements-dev.txt`. Or at least get the correct version if they do it piecemeal (btw, Having a robust packaging system instead of talking about this would be convenient.). I notice that `reno` is pinned, but there is an instruction to just `pip` install it.

Reply:
I tried a `ruff` target in `tox.ini`. It runs in 900ms. Running `ruff` directly is 23ms. And less if you run it only on the directory you are editing. The story with `black` is similar... But don't think the comments I added there are optimal. I mean giving explicit instructions for one invocation of `ruff` at the cli doesn't really belong there.